File: src/services/vmManager.js
import compute from '@google-cloud/compute';
import { config } from '../config.js';
import { v4 as uuidv4 } from 'uuid';

const instancesClient = new compute.InstancesClient({
  projectId: config.gcp.projectId,
  keyFilename: config.gcp.serviceAccountKeyPath
});

/**
 * 建立 Preemptible VM 實例
 * @param {string} instanceName - VM 實例名稱
 * @param {string} sgfGcsPath - SGF 檔案在 GCS 的路徑
 * @returns {Promise<Object>} VM 實例資訊
 */
export async function createPreemptibleVM(instanceName, sgfGcsPath) {
  // 建立啟動腳本（從 GCS 下載 SGF 並執行分析）
  const startupScript = `
#!/bin/bash
set -e

# 安裝 gcloud CLI（如果尚未安裝）
if ! command -v gsutil &> /dev/null; then
  echo "Installing gcloud CLI..."
  curl https://sdk.cloud.google.com | bash
  exec -l $SHELL
fi

# 下載 SGF 檔案
echo "Downloading SGF file from ${sgfGcsPath}..."
gsutil cp ${sgfGcsPath} /tmp/input.sgf

# 執行 KataGo 分析
echo "Running KataGo analysis..."
${config.katago.scriptPath} /tmp/input.sgf /tmp/result.txt

# 上傳結果至 GCS
echo "Uploading results..."
gsutil cp /tmp/result.txt gs://${config.storage.bucketName}/results/${instanceName}/result.txt

# 標記任務完成
echo "Task completed" > /tmp/status.txt
gsutil cp /tmp/status.txt gs://${config.storage.bucketName}/results/${instanceName}/status.txt

# 關閉 VM（分析完成後自動關閉）
sudo shutdown -h +1
`;

  const instanceResource = {
    name: instanceName,
    machineType: `zones/${config.gcp.zone}/machineTypes/${config.vm.machineType}`,
    disks: [
      {
        boot: true,
        autoDelete: true,
        initializeParams: {
          sourceImage: `projects/${config.vm.imageProject}/global/images/family/${config.vm.imageFamily}`,
          diskSizeGb: config.vm.diskSizeGB
        }
      }
    ],
    metadata: {
      items: [
        {
          key: 'startup-script',
          value: startupScript
        }
      ]
    },
    scheduling: {
      preemptible: config.vm.preemptible,
      onHostMaintenance: 'TERMINATE',
      automaticRestart: false
    },
    serviceAccounts: [
      {
        email: 'default',
        scopes: [
          'https://www.googleapis.com/auth/cloud-platform',
          'https://www.googleapis.com/auth/devstorage.read_write'
        ]
      }
    ],
    tags: {
      items: ['katago-worker']
    },
    networkInterfaces: [
      {
        network: 'global/networks/default',
        accessConfigs: [
          {
            name: 'External NAT',
            type: 'ONE_TO_ONE_NAT'
          }
        ]
      }
    ]
  };

  const [response] = await instancesClient.insert({
    project: config.gcp.projectId,
    zone: config.gcp.zone,
    instanceResource
  });

  // 等待操作完成
  const operationsClient = new compute.ZoneOperationsClient({
    projectId: config.gcp.projectId,
    keyFilename: config.gcp.serviceAccountKeyPath
  });

  let operation = response.latestResponse;
  while (operation.status !== 'DONE') {
    [operation] = await operationsClient.wait({
      operation: operation.name,
      project: config.gcp.projectId,
      zone: config.gcp.zone
    });
  }

  return {
    name: instanceName,
    zone: config.gcp.zone,
    status: 'PROVISIONING',
    sgfPath: sgfGcsPath
  };
}

/**
 * 取得 VM 狀態
 * @param {string} instanceName - VM 實例名稱
 * @returns {Promise<Object>} VM 狀態資訊
 */
export async function getVMStatus(instanceName) {
  try {
    const [instance] = await instancesClient.get({
      project: config.gcp.projectId,
      zone: config.gcp.zone,
      instance: instanceName
    });

    return {
      exists: true,
      status: instance.status,
      name: instance.name,
      zone: config.gcp.zone
    };
  } catch (error) {
    if (error.code === 404 || error.code === 5) {
      return { exists: false };
    }
    throw error;
  }
}

/**
 * 刪除 VM 實例
 * @param {string} instanceName - VM 實例名稱
 * @returns {Promise<void>}
 */
export async function deleteVM(instanceName) {
  try {
    const [response] = await instancesClient.delete({
      project: config.gcp.projectId,
      zone: config.gcp.zone,
      instance: instanceName
    });

    // 等待操作完成
    const operationsClient = new compute.ZoneOperationsClient({
      projectId: config.gcp.projectId,
      keyFilename: config.gcp.serviceAccountKeyPath
    });

    let operation = response.latestResponse;
    while (operation.status !== 'DONE') {
      [operation] = await operationsClient.wait({
        operation: operation.name,
        project: config.gcp.projectId,
        zone: config.gcp.zone
      });
    }
  } catch (error) {
    // 如果 VM 不存在，忽略錯誤
    if (error.code !== 404 && error.code !== 5) {
      throw error;
    }
  }
}

/**
 * 產生唯一的 VM 實例名稱
 * @returns {string}
 */
export function generateInstanceName() {
  const timestamp = Date.now();
  const uuid = uuidv4().split('-')[0];
  return `${config.vm.instanceNamePrefix}-${timestamp}-${uuid}`;
}

/**
 * 等待 VM 啟動完成
 * @param {string} instanceName - VM 實例名稱
 * @param {number} maxWaitTime - 最大等待時間（毫秒）
 * @returns {Promise<boolean>} 是否成功啟動
 */
export async function waitForVMReady(instanceName, maxWaitTime = 60000) {
  const startTime = Date.now();

  while (Date.now() - startTime < maxWaitTime) {
    const status = await getVMStatus(instanceName);

    if (status.exists && status.status === 'RUNNING') {
      return true;
    }

    if (status.exists && status.status === 'TERMINATED') {
      return false;
    }

    // 等待 2 秒後再檢查
    await new Promise((resolve) => setTimeout(resolve, 2000));
  }

  return false;
}

File: src/services/vm_manager.py
from google.cloud import compute_v1
from config import config
import uuid
from typing import Dict, Optional


def generate_instance_name() -> str:
    """Generate unique VM instance name"""
    timestamp = int(time.time() * 1000)
    uuid_str = str(uuid.uuid4()).split("-")[0]
    return f"{config['vm']['instance_name_prefix']}-{timestamp}-{uuid_str}"


async def create_preemptible_vm(instance_name: str, sgf_gcs_path: str) -> Dict[str, str]:
    """Create Preemptible VM instance"""
    # Create startup script (download SGF from GCS and execute analysis)
    startup_script = f"""#!/bin/bash
set -e

# Install gcloud CLI (if not already installed)
if ! command -v gsutil &> /dev/null; then
  echo "Installing gcloud CLI..."
  curl https://sdk.cloud.google.com | bash
  exec -l $SHELL
fi

# Download SGF file
echo "Downloading SGF file from {sgf_gcs_path}..."
gsutil cp {sgf_gcs_path} /tmp/input.sgf

# Execute KataGo analysis
echo "Running KataGo analysis..."
{config['katago']['script_path']} /tmp/input.sgf /tmp/result.txt

# Upload results to GCS
echo "Uploading results..."
gsutil cp /tmp/result.txt gs://{config['storage']['bucket_name']}/results/{instance_name}/result.txt

# Mark task completed
echo "Task completed" > /tmp/status.txt
gsutil cp /tmp/status.txt gs://{config['storage']['bucket_name']}/results/{instance_name}/status.txt

# Shutdown VM (auto shutdown after analysis completes)
sudo shutdown -h +1
"""
    
    # Create instance
    instance_client = compute_v1.InstancesClient()
    
    instance = compute_v1.Instance()
    instance.name = instance_name
    instance.machine_type = f"zones/{config['gcp']['zone']}/machineTypes/{config['vm']['machine_type']}"
    
    # Boot disk
    disk = compute_v1.AttachedDisk()
    disk.boot = True
    disk.auto_delete = True
    disk.initialize_params = compute_v1.AttachedDiskInitializeParams()
    disk.initialize_params.source_image = (
        f"projects/{config['vm']['image_project']}/global/images/family/{config['vm']['image_family']}"
    )
    disk.initialize_params.disk_size_gb = config["vm"]["disk_size_gb"]
    instance.disks = [disk]
    
    # Metadata
    instance.metadata = compute_v1.Metadata()
    instance.metadata.items = [
        compute_v1.Items(key="startup-script", value=startup_script)
    ]
    
    # Scheduling
    instance.scheduling = compute_v1.Scheduling()
    instance.scheduling.preemptible = config["vm"]["preemptible"]
    instance.scheduling.on_host_maintenance = "TERMINATE"
    instance.scheduling.automatic_restart = False
    
    # Service account
    instance.service_accounts = [
        compute_v1.ServiceAccount(
            email="default",
            scopes=[
                "https://www.googleapis.com/auth/cloud-platform",
                "https://www.googleapis.com/auth/devstorage.read_write"
            ]
        )
    ]
    
    # Network
    network_interface = compute_v1.NetworkInterface()
    network_interface.network = "global/networks/default"
    access_config = compute_v1.AccessConfig()
    access_config.name = "External NAT"
    access_config.type_ = "ONE_TO_ONE_NAT"
    network_interface.access_configs = [access_config]
    instance.network_interfaces = [network_interface]
    
    # Tags
    instance.tags = compute_v1.Tags()
    instance.tags.items = ["katago-worker"]
    
    # Insert instance
    operation = instance_client.insert(
        project=config["gcp"]["project_id"],
        zone=config["gcp"]["zone"],
        instance_resource=instance
    )
    
    # Wait for operation to complete
    operation_client = compute_v1.ZoneOperationsClient()
    while operation.status != compute_v1.Operation.Status.DONE:
        operation = operation_client.get(
            project=config["gcp"]["project_id"],
            zone=config["gcp"]["zone"],
            operation=operation.name
        )
        await asyncio.sleep(1)
    
    return {
        "name": instance_name,
        "zone": config["gcp"]["zone"],
        "status": "PROVISIONING",
        "sgfPath": sgf_gcs_path
    }


async def get_vm_status(instance_name: str) -> Dict[str, any]:
    """Get VM status"""
    try:
        instance_client = compute_v1.InstancesClient()
        instance = instance_client.get(
            project=config["gcp"]["project_id"],
            zone=config["gcp"]["zone"],
            instance=instance_name
        )
        
        return {
            "exists": True,
            "status": instance.status,
            "name": instance.name,
            "zone": config["gcp"]["zone"]
        }
    except Exception as error:
        if hasattr(error, "code") and error.code in [404, 5]:
            return {"exists": False}
        raise


async def delete_vm(instance_name: str):
    """Delete VM instance"""
    try:
        instance_client = compute_v1.InstancesClient()
        operation = instance_client.delete(
            project=config["gcp"]["project_id"],
            zone=config["gcp"]["zone"],
            instance=instance_name
        )
        
        # Wait for operation to complete
        operation_client = compute_v1.ZoneOperationsClient()
        while operation.status != compute_v1.Operation.Status.DONE:
            operation = operation_client.get(
                project=config["gcp"]["project_id"],
                zone=config["gcp"]["zone"],
                operation=operation.name
            )
            await asyncio.sleep(1)
    except Exception as error:
        # If VM doesn't exist, ignore error
        if not (hasattr(error, "code") and error.code in [404, 5]):
            raise


async def wait_for_vm_ready(instance_name: str, max_wait_time: int = 60000) -> bool:
    """Wait for VM to be ready"""
    start_time = time.time() * 1000
    
    while (time.time() * 1000 - start_time) < max_wait_time:
        status = await get_vm_status(instance_name)
        
        if status.get("exists") and status.get("status") == "RUNNING":
            return True
        
        if status.get("exists") and status.get("status") == "TERMINATED":
            return False
        
        # Wait 2 seconds before checking again
        await asyncio.sleep(2)
    
    return False


# Add missing imports
import time
import asyncio

